catalogd/internal/storage/localdir.go

```diff
@@ -2,113 +2,316 @@ package storage
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !allowedMethodSet.Has(r.Method) {
+				http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+				return
+			}
+			next.ServeHTTP(w, r)
+		})
+	}
+	return allowedMethodsHandler(mux, http.MethodGet, http.MethodHead)
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	w.Header().Add("Content-Type", "application/jsonl")
+	http.ServeContent(w, r, "", catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
+	w.Header().Set("Last-Modified", catalogStat.ModTime().UTC().Format(TimeFormat))
+	if checkPreconditions(w, r, catalogStat.ModTime()) {
+		w.WriteHeader(http.StatusNotModified)
+		return
+	}
+
+	schema := r.URL.Query().Get("schema")
+	pkg := r.URL.Query().Get("package")
+	name := r.URL.Query().Get("name")
+
+	if schema == "" && pkg == "" && name == "" {
+		// If no parameters are provided, return the entire catalog (this is the same as /api/v1/all)
 		w.Header().Add("Content-Type", "application/jsonl")
-		gzHandler.ServeHTTP(w, r)
-	})
-	mux.Handle(s.RootURL.Path, typeHandler)
-	return mux
+		http.ServeContent(w, r, "", catalogStat.ModTime(), catalogFile)
+		return
+	}
+	idx, err := s.getIndex(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	indexReader := idx.Get(catalogFile, schema, pkg, name)
+	serveJSONLines(w, r, indexReader)
 }
 
-func (s LocalDirV1) ContentExists(catalog string) bool {
-	file, err := os.Stat(filepath.Join(s.RootDir, catalog, v1ApiPath, v1ApiData))
+func (s *LocalDirV1) catalogData(catalog string) (*os.File, os.FileInfo, error) {
+	catalogFile, err := os.Open(catalogFilePath(s.catalogDir(catalog)))
 	if err != nil {
-		return false
+		return nil, nil, err
 	}
-	if !file.Mode().IsRegular() {
-		// path is not valid content
-		return false
+	catalogFileStat, err := catalogFile.Stat()
+	if err != nil {
+		return nil, nil, err
 	}
-	return true
+	return catalogFile, catalogFileStat, nil
 }
 
-// filesOnlyFilesystem is a file system that can open only regular
-// files from the underlying filesystem. All other file types result
-// in os.ErrNotExists
-type filesOnlyFilesystem struct {
-	FS fs.FS
+func httpError(w http.ResponseWriter, err error) {
+	var code int
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		code = http.StatusNotFound
+	case errors.Is(err, fs.ErrPermission):
+		code = http.StatusForbidden
+	case errors.Is(err, errInvalidParams):
+		code = http.StatusBadRequest
```

Comment:
I don't think I see any where that we call `httpError` with `errInvalidParams`. Seems we should add a check in the query handler for unexpected parameters?